apps/api_web/lib/api_web/rate_limiter/rate_limiter_concurrent.ex

```diff
@@ -0,0 +1,175 @@
+          if user.static_concurrent_limit >= 0,
+            do:
+              max(
+                user.static_concurrent_limit || 0,
+                Keyword.fetch!(
+                  Application.get_env(:api_web, :rate_limiter_concurrent),
+                  :max_registered_static
+                )
+              ),
+            else: user.static_concurrent_limit
+
+        {true, :anon} ->
+          Keyword.fetch!(
+            Application.get_env(:api_web, :rate_limiter_concurrent),
+            :max_anon_streaming
+          )
+
+        {false, :anon} ->
+          Keyword.fetch!(
+            Application.get_env(:api_web, :rate_limiter_concurrent),
+            :max_anon_static
+          )
+      end
+
+    remaining = limit - active_connections
+    at_limit? = remaining <= 0
+    {at_limit?, remaining, limit}
+  end
+
+  def add_lock(%ApiWeb.User{} = user, pid, event_stream?) do
+    if enabled?() do
+      {_type, key} = lookup(user, event_stream?)
+      pid_key = pid |> :erlang.pid_to_list() |> to_string
+      timestamp = get_current_unix_ts()
+
+      Logger.info(
+        "#{__MODULE__} event=add_lock user=#{inspect(user)} pid_key=#{pid_key} key=#{key} timestamp=#{timestamp}"
+      )
+
+      locks = user |> get_locks(event_stream?) |> Map.put(pid_key, timestamp)
+
+      Logger.info(
+        "#{__MODULE__} event=add_lock_after user=#{inspect(user)} pid_key=#{pid_key} key=#{key} timestamp=#{timestamp} locks=#{inspect(locks)}"
+      )
+
+      GenServer.call(__MODULE__, {:memcache_set, key, locks})
+    end
+
+    nil
+  end
+
+  def remove_lock(
+        %ApiWeb.User{} = user,
+        pid,
+        event_stream?,
+        pid_key \\ nil
+      ) do
+    if enabled?() do
+      {_type, key} = lookup(user, event_stream?)
+      pid_key = if pid_key, do: pid_key, else: pid |> :erlang.pid_to_list() |> to_string
+      locks_before = get_locks(user, event_stream?)
+      locks = locks_before |> Map.delete(pid_key)
+
+      Logger.info(
+        "#{__MODULE__} event=remove_lock user_id=#{user.id} pid_key=#{pid_key} key=#{key}"
+      )
+
+      GenServer.call(__MODULE__, {:memcache_set, key, locks})
+
+      Logger.info(
+        "#{__MODULE__} event=remove_lock_after user_id=#{user.id} pid_key=#{pid_key} key=#{key} locks=#{inspect(locks)}"
+      )
+    end
+
+    nil
+  end
+
+  def enabled? do
+    Keyword.fetch!(Application.get_env(:api_web, :rate_limiter_concurrent), :enabled) == True
+  end
+
+  def memcache? do
+    Keyword.fetch!(Application.get_env(:api_web, :rate_limiter_concurrent), :memcache) == True
+  end
+
+  def handle_call({:memcache_get, key, default_value}, _from, state) do
+    {:reply,
+     {:ok,
+      case Memcache.get(state.memcache_pid, key) do
+        {:ok, result} -> result
+        _ -> default_value
+      end}, state}
+  end
+
+  def handle_call({:memcache_set, key, value}, _from, state) do
+    {:reply, {:ok, Memcache.set(state.memcache_pid, key, value)}, state}
```

Comment:
* Yes - I believe this could only be fixed by using a data store that understands transactions (such as redis). In practice, I suspect that the operation will be fast enough that we won't run into this, and I feel it's worth at least testing. 

* I pushed an update to always include an instance specific UUID as well to cover this, thanks!

Reply:
The UUID fix has been pushed - somehow I broke a merge from the default branch, so that's why I took this back to draft but it should be correct now.

Reply:
I have updated to use `cas` - the memcache library provides a clean function where you can supply an anonymous function that updates the value w/ automatic retry.